File: leetcode/Binary_Tree_Level_Order_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        if(root == nullptr) return {};
        if(!root->left && !root->right) return {{root->val}};
        
        vector<vector<int>> result;
        queue<TreeNode *> p, q;
        p.push(root);
        
        vector<int> layer;
        while(!p.empty()){
            TreeNode *t = p.front();
            p.pop();
            
            layer.push_back(t->val);
            if(t->left) q.push(t->left);
            if(t->right) q.push(t->right);
            
            if(p.empty()){
                swap(p, q);
                result.push_back(layer);
                layer.clear();
            }
        }
        
        return result;
    }
};

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        if(!root) return {};
        if(!root->left && !root->right) return {{root->val}};
        
        queue<TreeNode *> que;
        que.push(root);
        que.push(nullptr);
        
        vector<vector<int>> result;
        vector<int> layer;

        while(!que.empty()){
            TreeNode *t = que.front();
            que.pop();
            if(t == nullptr){
                // no more elems to traverse                 
                if(layer.empty()) break;

                result.push_back(layer);
                que.push(nullptr);
                layer.clear();
            }else{
                layer.push_back(t->val);
                if(t->left) que.push(t->left);
                if(t->right) que.push(t->right);
            }
        }
        
        return result;
    }
};

File: leetcode/java/BinaryTreeLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/* Solution #1 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root == null) return Collections.emptyList();
        
        TreeNode d = new TreeNode(1);
        Queue<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        q.offer(d);
        
        List<Integer> layer = new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        while(!q.isEmpty()) {
            TreeNode t = q.poll();
            if(t != d) {
                layer.add(t.val);
                if(t.left != null) q.offer(t.left);
                if(t.right != null) q.offer(t.right);
            }else{
                if(!q.isEmpty()) q.offer(d);
                result.add(layer);
                layer = new ArrayList<>();
            }
        }
        return result;
    }
}

/* Solution #2 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root == null) return Collections.emptyList();
        Queue<TreeNode> p = new ArrayDeque<>();
        Queue<TreeNode> q = new ArrayDeque<>();
        p.offer(root);

        List<List<Integer>> result = new ArrayList<>();
        List<Integer> layer = new ArrayList<>();
        while(!p.isEmpty()){
            TreeNode t = p.poll();
            layer.add(t.val);
            if(t.left != null) q.offer(t.left);
            if(t.right != null) q.offer(t.right);
            if(p.isEmpty()) {
                Queue<TreeNode> k = p;
                p = q;
                q = k;
                result.add(layer);
                layer = new ArrayList<>();
            }
        }
        return result;
    }
}
